File: 221.maximal-square.cpp
/*
 * @lc app=leetcode id=221 lang=cpp
 *
 * [221] Maximal Square
 */

// @lc code=start
#include <vector>
using namespace std;
class Solution {
 private:
  int maximalSquareDP1(vector<vector<char>>& matrix) {
    if (matrix.empty()) return 0;
    int m = matrix.size(), n = matrix[0].size();
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));
    int maxSquareLen = 0;
    for (int i = 0; i < m; i++) {
      for (int j = 0; j < n; j++) {
        if (matrix[i][j] == '1') {
          dp[i + 1][j + 1] = min(min(dp[i][j + 1], dp[i + 1][j]), dp[i][j]) + 1;
        }
        maxSquareLen = max(maxSquareLen, dp[i + 1][j + 1]);
      }
    }
    return maxSquareLen * maxSquareLen;
  }
  int maximalSquareDP2(vector<vector<char>>& matrix) {
    if (!matrix.size()) return 0;
    const int rows = matrix.size(), columns = matrix[0].size();
    vector<vector<int>> dp(2, vector<int>(columns + 1, 0));
    int longestSide = 0, previousRow = 0, currentRow = 1;
    for (int r = 1; r <= rows; r++) {
      currentRow = r & 1;
      for (int c = 1; c <= columns; c++) {
        if (matrix[r - 1][c - 1] == '0') {
          dp[currentRow][c] = 0;
        } else {
          int side = 1 + min(dp[previousRow][c - 1], min(dp[previousRow][c], dp[currentRow][c - 1]));
          longestSide = max(longestSide, side);
          dp[currentRow][c] = side;
        }
      }
      previousRow = currentRow;
    }
    return longestSide * longestSide;
  }

 public:
  int maximalSquare(vector<vector<char>>& matrix) {}
};
// @lc code=end

File: 221.maximal-square.java
/*
 * @lc app=leetcode id=221 lang=java
 *
 * [221] Maximal Square
 */

// @lc code=start
class Solution {
  public int maximalSquare(char[][] matrix) {
    if (matrix.length == 0)
      return 0;
    int maxEdge = 0;
    int[][] dp = new int[2][matrix[0].length];

    for (int i = 0; i < matrix.length; i++) {
      for (int j = 0; j < matrix[0].length; j++) {
        if (i == 0 || j == 0)
          dp[1][j] = Character.getNumericValue(matrix[i][j]);
        else if (matrix[i][j] == '1') {
          int min = Math.min(dp[0][j], Math.min(dp[0][j - 1], dp[1][j - 1]));
          dp[1][j] = min + 1;
        }
        maxEdge = Math.max(dp[1][j], maxEdge);
      }
      for (int k = 0; k < dp[0].length; k++) {
        dp[0][k] = dp[1][k];
        dp[1][k] = 0;
      }
    }
    return maxEdge * maxEdge;
  }
}
// @lc code=end

